### PR TITLE
feat: namespace files support (open, upload, sync)

### LINKED ISSUE #47: Namespace Files support — edit and deploy from the editor
Thank you again for quick fixes that I described in earlier issues.

Now this is maybe more of a feature request and I am not sure if this is in your roadmap/plan for this extension, but it would be very helpful.
It was already mentioned in https://github.com/kestra-io/vscode-kestra/pull/11#pullrequestreview-1740887926 but as I am aware, there were no follow-ups.

The extension handles flow YAML end to end (validation, save, run, topology). Anything else that lives on a namespace — Python scripts, SQL, config, ... — has to be managed either in the web UI or via `kestractl nsfiles` from the terminal (which works good). 

Would it make sense to have commands that will upload namespace files to Kestra instance, e.g.:
* **Kestra: Upload File to Namespace**
* **Kestra: Sync Folder to Namespace**

As mentioned, as an alternative we are currently using `kestractl nsfiles upload/list/get` commands in CLI, but as extension already has option to save flows to Kestra instance, it would be very useful to save the namespace files, which these flows are using, also with extension.

### PR DESCRIPTION
**Pull request by Malaydewangan09**
- closes #47

### CODE REVIEW

**Comment by Malaydewangan09** [comment]
Screenshot 2026-07-30 at 1 10 37 PM


Screenshot 2026-07-30 at 1 11 56 PM

File: src/web/apiClient.ts [hunk 1]
```diff
@@ -273,6 +278,71 @@ export default class ApiClient {
         return response?.ok ? (await response.json().catch(() => null)) as FlowGraph | null : null;
     }
 
+    // Existing namespaces on the instance, to populate the "Open namespace" picker. Paged through in
+    // full so instances with more than one page are not silently truncated.
+    public async listNamespaces(): Promise<string[]> {
+        const size = 200;
```
Comment (jymaire) [inline src/web/apiClient.ts:285]: what happens if customer has 201 namespaces? (maybe 200 is already too many, but just wondering with this number)

File: src/web/apiClient.ts [hunk 2]
```diff
@@ -273,6 +278,71 @@ export default class ApiClient {
...
+        const ids: string[] = [];
+        for (let page = 1; page <= 50; page++) {
+            const response = await this.silentFetch(`/namespaces/search?existing=true&size=${size}&page=${page}&sort=id%3Aasc`);
+            if (!response?.ok) {
+                break;
```
Comment (jymaire) [inline src/web/apiClient.ts]: do we have some logs available in VScode extensions?

**Comment by jymaire** [comment]
Nice feature, and the pre-flight reachability check with a distinct message per failure mode (401 split on stored-credential presence, 403, 404, unreachable) is a real improvement over a silently empty window. The modal confirmation before a bulk sync, the cancellable progress, the batch-summary reporting instead of per-file toasts, and the documented symlink-cycle reasoning in `collectFiles` all show care. `KestraFS.start()` is strictly better than what it replaces.

A few things I'd like addressed:

### 1. Upload/sync to a namespace that doesn't exist yet is blocked

`apiClient.ts:331-333` rejects a namespace when `namespaceExists` returns `false`, and `resolveConfiguredNamespace` (`namespaceFiles.ts:72`) is on the path for **all three** commands. That gate makes sense for *open* — the comment explains it, a bad name would otherwise auto-create an empty dir. It's wrong for *upload* and *sync*, where creating files in a fresh namespace is a normal thing to want, and the code's own comment confirms the POST endpoint auto-creates.

This also makes the picker's "Enter namespace manually…" affordance (`namespaceFiles.ts:15`, commented "always allows typing a new one") dead for anything new: you type it, then get *"Namespace … was not found"*.

Suggest parameterizing the pre-flight — `resolveConfiguredNamespace(apiClient, {requireExisting: boolean})` — or, for upload/sync, downgrading non-existence to a confirmation prompt ("Namespace X does not exist yet. Create it?").

### 2. Uploads read from disk, ignoring unsaved editor changes

`namespaceFiles.ts:121` uses `vscode.workspace.fs.readFile(fileUri)`. With no `resource` argument the file comes from `activeTextEditor.document`, so a dirty buffer silently uploads the *stale* on-disk bytes and then reports success.

This diverges from the project's own convention: `flowRunner.ts`'s `activeFlow()` uses `document.getText()`. Either `await document.save()` first (matching the "Save and run flow" pattern) or prefer the open document's text when one matches the URI.

### 3. Explorer multi-select silently drops all but one file

VS Code invokes `explorer/context` commands as `(clickedUri, selectedUris[])`. Both handlers in `extension.ts` accept only the first parameter, so selecting 12 files and choosing "Upload file to namespace" uploads one. Accept the second argument and loop, or at minimum name the command so the singular is expected.

### 4. `uploadNamespaceFile` bypasses `fetchWithTimeout`

`apiClient.ts:323` uses bare `fetch`, while every other call in the class routes through `ApiClient.fetchWithTimeout` (15s) via `silentFetch`/`apiCall`.

Consequence during a folder sync: `pushFiles` only checks `token.isCancellationRequested` *between* files (`namespaceFiles.ts:170`), so one hung upload leaves the progress notification stuck forever with a Cancel button that does nothing. Route through `fetchWithTimeout` and pass the cancellation token's signal through.

### 5. Folder sync will push secrets by default

`IGNORED_NAMES` (`namespaceFiles.ts:5`) covers `.git`, `.vscode`, `node_modules`, `.DS_Store`, `.idea` — but not `.env`, `.env.local`, `*.pem`, `credentials.json`, `.npmrc`. Syncing a typical project root therefore uploads local secrets to a shared Kestra instance, and the confirm dialog only shows a file *count*, not what's included.

Either exclude `.env*` and common credential patterns, honour `.gitignore`, or show the file list (a QuickPick with everything pre-checked would let users deselect). This is the one I'd most want addressed before merge.

### 6. `collectFiles` has no error handling

`namespaceFiles.ts:135` — a `readDirectory` failure (permissions, a broken mount) propagates out of the command handler as an unhandled rejection, which surfaces as VS Code's generic "command failed" popup. Everything else in this file reports failures deliberately; this path should too. Consider also skipping unreadable subdirectories rather than aborting the whole sync.

### 7. Context-menu items always visible

The two `explorer/context` entries have no gating on whether a Kestra instance is configured, so they appear on every file and folder in every workspace for everyone who installs the extension. By contrast, the `editor/context` entries in this repo are all gated (`resourceLangId == yaml`). Add a `when` clause on the config being set, or nest them under a "Kestra" submenu.

Also, `group: "kestra@1"` is a custom group with no anchor, so ordering relative to other extensions is undefined — `navigation@…` or an explicit known group is more predictable.

### 8. No tests

`src/test/` has seven mocha suites for pure helpers, and this PR adds none. `namespacePath` (the `/` base-path edge case is exactly the kind of thing that regresses), `basename`, `reachabilityError`'s status→message mapping, and `reportSyncOutcome`'s branch selection are all pure and cheap to cover. Export them and add `src/test/namespaceFiles.test.ts`.

---

There's also a set of smaller points (unpaged `namespaceExists` vs. the carefully paged `listNamespaces`, an off-by-one in the progress message, `..` surviving `encodeURIComponent` in the target path, `IGNORED_NAMES` duplicating `EXCLUDED_FOLDERS`, sequential uploads) that I'm happy to write up separately if useful.

File: src/web/apiClient.ts (hunk 1 above)
Comment (Malaydewangan09) [reply src/web/apiClient.ts:285]: It fetches up tp 50 pages, 200 for each page, so its fine

**Comment by jymaire** [comment]
Thanks for the quick turnaround — `fa11223` addresses most of this well. Extracting the pure helpers into a vscode-free `namespaceFilesHelpers.ts` with real test coverage is exactly the right shape, and the multi-select handling (batch path for a selection, exact-path flow preserved for a single file) is better than what I suggested. Issues 1, 3, 6 and 8 are fully resolved; CI is green.

Two things I'd like fixed before approving, plus a few smaller ones.

## Blocking

### A. `syncFolderToNamespace` always opens the folder picker — `namespaceFiles.ts:280`

```ts
const uris = resolveUris(resource, selected, await pickLocalFolder());
```

Argument expressions evaluate before the call, so `pickLocalFolder()` runs unconditionally. Right-clicking a folder in the Explorer now pops an unwanted "Sync folder" open-dialog, and whatever the user picks is discarded in favour of `resource`. The pre-refactor `resource ?? await pickLocalFolder()` short-circuited correctly.

Needs a lazy fallback — pass a thunk, or resolve the picker only when `resource` and `selected` are both empty.

(The `uploadFileToNamespace` call site is fine; its fallback is a plain property read with no side effect.)

### B. `kestra.hasApiUrl` is set too late to matter — `extension.ts:84,122`

`activationEvents` is `["onFileSystem:kestra", "onLanguage:yaml"]` plus the implicit `onCommand` entries. The context key is only set inside `activate()`, so in a workspace with no YAML file open and no `kestra:` folder, the extension hasn't activated, the key is undefined, and the menu items **never appear** — even for users who have configured a URL. Right-clicking a `.py` file in a fresh window shows nothing until something unrelated activates the extension.

Cleanest fix is to drop the context key entirely and use VS Code's native config check in the `when` clause, which needs no activation and no plumbing:

```json
"when": "config.kestra.api.url && resourceScheme == file && !explorerResourceIsFolder"
```

Otherwise, add `onStartupFinished` to `activationEvents`.

## Worth fixing, not blocking

### C. `readLocalFile` re-encodes clean documents — `namespaceFiles.ts:77-80`

The lookup matches any open document, not just dirty ones, so a *clean* open file gets `getText()` re-encoded to UTF-8 instead of its disk bytes. For a file with a UTF-8 BOM, or under a non-UTF-8 `files.encoding`, the uploaded bytes then differ from what's on disk. Gating on `open.isDirty` gets the unsaved-content fix without changing behaviour for clean files.

### D. Ignore list doesn't apply to explicit selections — `namespaceFiles.ts:124`

`isIgnoredName` filters the recursive walk, but `gatherFiles` pushes a directly-selected file unchecked. Select-all in the Explorer → "Upload file to namespace" uploads `.env` anyway, which is the case the secret-exclusion work was aimed at. Treating an explicit pick as intent is defensible for a *single* file; for a multi-selection it's more likely a blanket selection. Suggest filtering multi-selections and surfacing what was skipped.

### E. Single-file upload has no timeout and no cancel — `namespaceFiles.ts:232`

`uploadNamespaceFile` still uses bare `fetch` rather than `ApiClient.fetchWithTimeout`, and `uploadSingleFile` passes no signal and shows no progress UI. A large or stalled single-file upload hangs silently with no way out. The batch path is fine now thanks to the `AbortController` — routing `uploadNamespaceFile` through `fetchWithTimeout` would close the remaining gap for both.

### F. Minor

- Two selected folders sharing a basename collapse to the same namespace path in `gatherFiles`, last write wins silently.
- `.envrc`, `*.p8` and `.pypirc` aren't in the secret list. Diminishing returns, entirely optional.

---

A is a two-line fix and B is arguably a one-line `when`-clause change, so this should be quick. Happy to re-review as soon as those two are in.

**Review by jymaire — APPROVED** [review]
